File: web/src/api/genetics.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) OwnPulse Contributors

import { useAuthStore } from "../store/auth";
import { api } from "./client";

export interface GeneticRecord {
  rsid: string;
  chromosome: string;
  position: number;
  genotype: string;
  created_at: string;
}

export interface GeneticListResponse {
  records: GeneticRecord[];
  total: number;
  page: number;
  per_page: number;
}

export interface GeneticSummary {
  total_variants: number;
  source: string | null;
  uploaded_at: string | null;
  chromosomes: Record<string, number>;
  annotated_count: number;
}

export interface UploadResult {
  total_variants: number;
  new_variants: number;
  duplicates_skipped: number;
  format: string;
  source: string;
}

export interface Interpretation {
  rsid: string;
  gene: string | null;
  chromosome: string;
  position: number;
  user_genotype: string;
  category: "health_risk" | "trait" | "pharmacogenomics" | "carrier_status";
  title: string;
  summary: string;
  risk_level:
    | "high"
    | "moderate"
    | "low"
    | "normal"
    | "poor_metabolizer"
    | "intermediate"
    | "rapid";
  significance: string;
  evidence_level: "strong" | "moderate" | "limited" | "preliminary";
  source: string;
  source_id: string | null;
  population_frequency: number | null;
  details: Record<string, unknown>;
}

export interface InterpretationsResponse {
  interpretations: Interpretation[];
  disclaimer: string;
}

/**
 * Upload genetic data file (23andMe, AncestryDNA, VCF).
 * Uses fetch directly because the shared api client forces JSON Content-Type,
 * which is incompatible with FormData/multipart uploads.
 * Auth token is read from the Zustand auth store (in-memory only).
 */
async function uploadGeneticFile(file: File): Promise<UploadResult> {
  const token = useAuthStore.getState().token;
  const formData = new FormData();
  formData.append("file", file);

  const headers: Record<string, string> = {};
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  const response = await fetch("/api/v1/genetics/upload", {
    method: "POST",
    headers,
    body: formData,
    credentials: "include",
  });

  if (response.status === 401) {
    useAuthStore.getState().logout();
    throw new Error("Unauthorized");
  }

  if (!response.ok) {
    const body = await response.text();
    throw new Error(body);
  }

  return response.json() as Promise<UploadResult>;
}

export const geneticsApi = {
  upload: uploadGeneticFile,

  list: (params?: { page?: number; per_page?: number; chromosome?: string; rsid?: string }) => {
    const searchParams = new URLSearchParams();
    if (params?.page) searchParams.set("page", String(params.page));
    if (params?.per_page) searchParams.set("per_page", String(params.per_page));
    if (params?.chromosome) searchParams.set("chromosome", params.chromosome);
    if (params?.rsid) searchParams.set("rsid", params.rsid);
    const qs = searchParams.toString();
    return api.get<GeneticListResponse>(`/api/v1/genetics${qs ? `?${qs}` : ""}`);
  },

  summary: () => api.get<GeneticSummary>("/api/v1/genetics/summary"),

  interpretations: (category?: string) => {
    const url = category
      ? `/api/v1/genetics/interpretations?category=${category}`
      : "/api/v1/genetics/interpretations";
    return api.get<InterpretationsResponse>(url);
  },

  deleteAll: () => api.delete<void>("/api/v1/genetics"),
};

File: backend/api/src/models/genetics.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) OwnPulse Contributors

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// A row from the `genetic_records` table.
#[derive(FromRow, Serialize)]
pub struct GeneticRecordRow {
    pub id: Uuid,
    pub user_id: Uuid,
    pub source: String,
    pub rsid: Option<String>,
    pub chromosome: Option<String>,
    pub position: Option<i64>,
    pub genotype: Option<String>,
    pub uploaded_file_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

/// Result of a genetic data upload.
#[derive(Serialize)]
pub struct UploadResult {
    pub total_variants: i64,
    pub new_variants: i64,
    pub duplicates_skipped: i64,
    pub format: String,
    pub source: String,
}

/// Query params for listing genetic records.
#[derive(Deserialize)]
pub struct GeneticsListQuery {
    #[serde(default = "default_page")]
    pub page: i64,
    #[serde(default = "default_per_page")]
    pub per_page: i64,
    pub chromosome: Option<String>,
    pub rsid: Option<String>,
}

fn default_page() -> i64 {
    1
}

fn default_per_page() -> i64 {
    50
}

/// Response for the genetics list endpoint.
#[derive(Serialize)]
pub struct GeneticsListResponse {
    pub records: Vec<GeneticRecordRow>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

/// Summary of a user's genetic data.
#[derive(Serialize)]
pub struct GeneticSummary {
    pub total_variants: i64,
    pub source: Option<String>,
    pub uploaded_at: Option<DateTime<Utc>>,
    pub chromosomes: std::collections::HashMap<String, i64>,
    pub annotated_count: i64,
}

/// Query params for the interpretations endpoint.
#[derive(Deserialize)]
pub struct InterpretationsQuery {
    pub category: Option<String>,
}

/// A single interpreted variant — the user's genotype joined with annotation data.
#[derive(FromRow, Serialize, Clone)]
pub struct InterpretationRow {
    pub rsid: String,
    pub gene: Option<String>,
    pub chromosome: Option<String>,
    pub position: Option<i64>,
    pub user_genotype: Option<String>,
    pub category: String,
    pub title: String,
    pub summary: String,
    pub risk_allele: Option<String>,
    pub normal_allele: Option<String>,
    pub significance: String,
    pub evidence_level: String,
    pub source: String,
    pub source_id: Option<String>,
    pub population_frequency: Option<f64>,
    pub details: serde_json::Value,
}

/// A fully interpreted variant with computed risk level and personalized summary.
#[derive(Serialize)]
pub struct Interpretation {
    pub rsid: String,
    pub gene: Option<String>,
    pub chromosome: Option<String>,
    pub position: Option<i64>,
    pub user_genotype: Option<String>,
    pub category: String,
    pub title: String,
    pub summary: String,
    pub risk_level: String,
    pub significance: String,
    pub evidence_level: String,
    pub source: String,
    pub source_id: Option<String>,
    pub population_frequency: Option<f64>,
    pub details: serde_json::Value,
}

/// Response for the interpretations endpoint.
#[derive(Serialize)]
pub struct InterpretationsResponse {
    pub interpretations: Vec<Interpretation>,
    pub disclaimer: String,
}

/// Request body for delete confirmation.
#[derive(Deserialize)]
pub struct DeleteConfirmation {
    pub confirm: bool,
}

/// DB row for chromosome count aggregation.
#[derive(FromRow)]
pub struct ChromosomeCount {
    pub chromosome: Option<String>,
    pub count: Option<i64>,
}

/// DB row for summary source info.
#[derive(FromRow)]
pub struct SourceInfo {
    pub source: Option<String>,
    pub uploaded_at: Option<DateTime<Utc>>,
}
